sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/BuilderHelper.java

```diff
@@ -45,27 +46,33 @@
     static HttpPipeline buildPipeline(
         TablesSharedKeyCredential tablesSharedKeyCredential,
         TokenCredential tokenCredential, AzureSasCredential azureSasCredential, String sasToken,
-        String endpoint, RequestRetryOptions retryOptions, HttpLogOptions logOptions,
-        HttpClient httpClient, List<HttpPipelinePolicy> additionalPolicies,
-        Configuration configuration, ClientLogger logger) {
+        String endpoint, RequestRetryOptions retryOptions, HttpLogOptions logOptions, ClientOptions clientOptions,
+        HttpClient httpClient, List<HttpPipelinePolicy> perCallAdditionalPolicies,
+        List<HttpPipelinePolicy> perRetryAdditionalPolicies, Configuration configuration, ClientLogger logger) {
+
+        configuration = (configuration == null) ? Configuration.getGlobalConfiguration() : configuration;
+        clientOptions = (clientOptions == null) ? new ClientOptions() : clientOptions;
+        logOptions = (logOptions == null) ? new HttpLogOptions() : logOptions;
 
         validateSingleCredentialIsPresent(
             tablesSharedKeyCredential, tokenCredential, azureSasCredential, sasToken, logger);
 
-        //1
+        // Closest to API goes first, closest to wire goes last.
         List<HttpPipelinePolicy> policies = new ArrayList<>();
-        policies.add(getUserAgentPolicy(configuration));
+        policies.add(getUserAgentPolicy(configuration, clientOptions, logOptions));
```

Comment:
Where are the `clientOptions.getHeaders()` used? Don't they need to be added to the AddHeadersPolicy?

Reply:
Thanks Srikanta, will add the headers from `ClientOptions` as well.